### PR TITLE
Add an impl for Box<Error> from String.

### LINKED ISSUE #30156: String, From and Box<Error> interoperation could be better.
There's an implementation of `From` that allows the conversion of a `String` to a `Box<Error + Send + Sync>`, but it doesn't quite work when just using `Box<Error>`.

## Code

```rust
use std::error::Error;

fn main() {
    // Ok.
    let e1: Box<Error + Send + Sync> = From::from("hello".to_string());
    let e2: Box<Error> = e1;

    // Not ok.
    let e3: Box<Error> = From::from("hello".to_string());

    println!("{:?} {:?}", e2, e3);
}
```

## Compiler Output

```
michael: ~ $ rustc -V
rustc 1.6.0-nightly (9303055f3 2015-11-19)

michael: ~ $ rustc test.rs 
test.rs:9:26: 9:36 error: the trait `std::error::Error` is not implemented for the type `collections::string::String` [E0277]
test.rs:9     let e3: Box<Error> = From::from("hello".to_string());
                                   ^~~~~~~~~~
test.rs:9:26: 9:36 help: run `rustc --explain E0277` to see a detailed explanation
test.rs:9:26: 9:36 note: required by `core::convert::From::from`
error: aborting due to previous error
```

The (perhaps subtle) error message is due to the compiler attempting to use the general `From<E: Error>` implementation for `Box<Error>`.

## Solutions and Workarounds

I was trying to return string errors from a function returning `Result<_, Box<Error>>` and ended up writing the following:

```rust
fn string_error<T>(s: String) -> Result<T, Box<Error>> {
    let err: Box<Error + Send + Sync> = From::from(s);
    Err(err as Box<Error>)
}
```

This works well, but is really just boilerplate and not obviously suited to inclusion in any crate or library. Other solutions I thought of include:

* Adding `impl From<String> for Box<Error>` to the standard library for cases like this.
* Altering the semantics for trait-resolution on types that parametrise over multiple traits. From what I've read, we don't yet have semantics for general additions of traits, only built-in ones. See [E0225](http://doc.rust-lang.org/error-index.html#E0225).

### PR DESCRIPTION
**Pull request by michaelsproul**
Closes #30156.

### CODE REVIEW

File: src/libstd/error.rs [hunk 1]
```diff
@@ -119,6 +119,15 @@ impl From<String> for Box<Error + Send + Sync> {
     }
 }
 
+#[unstable(feature = "string_box_error", reason = "recently added", issue = "30156")]
```
Comment (sfackler) [inline src/libstd/error.rs]: Implementations aren't checked for stability so this should really be flagged stable.

Comment (michaelsproul) [inline src/libstd/error.rs]: I was wondering about that. I'll amend the commit and do a push.

**Comment by michaelsproul** [comment]
@alexcrichton this seems to have stalled (no acknowledgement from @bors), and the r+ needs to be on the new commit. It's holidays though so take it easy! :smile:

**Comment by netvl** [comment]
Is there a reason why there is no similar implementation for `&str`?

**Comment by alexcrichton** [comment]
@michaelsproul oh if the commit is updated it needs a new r+

As @netvl says actually, could you add an impl for `&str` as well?

**Comment by michaelsproul** [comment]
Will fix it up soon.

**Comment by michaelsproul** [comment]
Updated! I really like the `&str` implementation, great idea @netvl.

**Comment by alexcrichton** [comment]
@bors: r+ c1e527f11dfecd57e11b72a5f77725295f4b5ca4

cc @rust-lang/libs, just a heads up, seems easy though as we have `From<&str> for Box<Error+Send+Sync>` so it seems like there's no reason to not also have it for `Box<Error>`

**Comment by brson** [comment]
bhsausabsha on IRC asked for a backport

**Comment by bluss** [comment]
backporting would solve this beta regression IIUC #30634

**Comment by brson** [comment]
Per @nikomatsakis accepted for backport. This fixes another regression.

**Pull request merged**